### PR TITLE
[inductor] use same method to handle exception with aten

### PR DESCRIPTION
**Pull request by zhuhaozhe**
Fix https://github.com/pytorch/pytorch/issues/126691.
There are two reasons lead to this issue:
 - First issue is inductor use different ways to handle c++ error. For other [torch pybinds](https://github.com/pytorch/pytorch/blob/2ad0e4197d1c6b2f0f732f2a6ec29de2af031c28/torch/csrc/README.md#files)/aten binding , they are all handled by [HANDLE_TH_ERRORS](https://github.com/pytorch/pytorch/blob/2ad0e4197d1c6b2f0f732f2a6ec29de2af031c28/torch/csrc/Exceptions.h#L49). One of the point of `HANDLE_TH_ERRORS` is it translate c++ error into python errors, and it is registered here https://github.com/pytorch/pytorch/blob/2ad0e4197d1c6b2f0f732f2a6ec29de2af031c28/torch/csrc/Module.cpp#L1711-L1718 for aten. 
With the translations, we will get similar behavior with aten ops:
```
Traceback (most recent call last):
  File "/localdisk/haozhe/pt-upstream/126691.py", line 23, in <module>
    fn(x, y)
  File "/localdisk/haozhe/pt-upstream/pytorch/torch/_dynamo/eval_frame.py", line 433, in _fn
    return fn(*args, **kwargs)
  File "/localdisk/haozhe/pt-upstream/126691.py", line 17, in fn
    @torch.compile
  File "/localdisk/haozhe/pt-upstream/pytorch/torch/_dynamo/eval_frame.py", line 600, in _fn
    return fn(*args, **kwargs)
  File "/localdisk/haozhe/pt-upstream/pytorch/torch/_functorch/aot_autograd.py", line 1004, in forward
    return compiled_fn(full_args)
  File "/localdisk/haozhe/pt-upstream/pytorch/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 217, in runtime_wrapper
    all_outs = call_func_at_runtime_with_args(
  File "/localdisk/haozhe/pt-upstream/pytorch/torch/_functorch/_aot_autograd/utils.py", line 120, in call_func_at_runtime_with_args
    out = normalize_as_list(f(args))
  File "/localdisk/haozhe/pt-upstream/pytorch/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 451, in wrapper
    return compiled_fn(runtime_args)
  File "/localdisk/haozhe/pt-upstream/pytorch/torch/_inductor/codecache.py", line 1158, in __call__
    return self.current_callable(inputs)
  File "/tmp/torchinductor_root/i4/ci43bb3jykci3cnxgtdbv2odixxkdkiiy4juvo3usq5vdi3issw7.py", line 65, in call
    cpp_fused_add_index_0(arg0_1, arg1_1, buf0)
RuntimeError: index out of bounds: tmp1 < 1024L
```
And without the translation
```
Traceback (most recent call last):
  File "/localdisk/haozhe/pt-upstream/126691.py", line 23, in <module>
    fn(x, y)
  File "/localdisk/haozhe/pt-upstream/pytorch/torch/_dynamo/eval_frame.py", line 433, in _fn
    return fn(*args, **kwargs)
  File "/localdisk/haozhe/pt-upstream/126691.py", line 17, in fn
    @torch.compile
  File "/localdisk/haozhe/pt-upstream/pytorch/torch/_dynamo/eval_frame.py", line 600, in _fn
    return fn(*args, **kwargs)
  File "/localdisk/haozhe/pt-upstream/pytorch/torch/_functorch/aot_autograd.py", line 1004, in forward
    return compiled_fn(full_args)
  File "/localdisk/haozhe/pt-upstream/pytorch/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 217, in runtime_wrapper
    all_outs = call_func_at_runtime_with_args(
  File "/localdisk/haozhe/pt-upstream/pytorch/torch/_functorch/_aot_autograd/utils.py", line 120, in call_func_at_runtime_with_args
    out = normalize_as_list(f(args))
  File "/localdisk/haozhe/pt-upstream/pytorch/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 451, in wrapper
    return compiled_fn(runtime_args)
  File "/localdisk/haozhe/pt-upstream/pytorch/torch/_inductor/codecache.py", line 1158, in __call__
    return self.current_callable(inputs)
  File "/tmp/torchinductor_root/i4/ci43bb3jykci3cnxgtdbv2odixxkdkiiy4juvo3usq5vdi3issw7.py", line 65, in call
    cpp_fused_add_index_0(arg0_1, arg1_1, buf0)
RuntimeError: index out of bounds: tmp1 < 1024L
Exception raised from kernel at /tmp/torchinductor_root/em/cemt4dfdm36tf54ulcrusaknrfn25l7lndm4usdusen7tofzgei4.cpp:15 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0xac (0x7f326384d97c in /localdisk/haozhe/pt-upstream/pytorch/torch/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, char const*) + 0x7d (0x7f32637f67a9 in /localdisk/haozhe/pt-upstream/pytorch/torch/lib/libc10.so)
frame #2: <unknown function> + 0x8596 (0x7f325e41e596 in /tmp/torchinductor_root/em/cemt4dfdm36tf54ulcrusaknrfn25l7lndm4usdusen7tofzgei4.so)
frame #3: <unknown function> + 0x85f9 (0x7f325e41e5f9 in /tmp/torchinductor_root/em/cemt4dfdm36tf54ulcrusaknrfn25l7lndm4usdusen7tofzgei4.so)
frame #4: python() [0x5073c0]
<omitting python frames>
frame #7: python() [0x4f80b3]
frame #9: python() [0x4f80b3]
frame #12: python() [0x5cb113]
frame #15: python() [0x4e69da]
frame #18: python() [0x4e69da]
frame #21: python() [0x4e69da]
frame #24: python() [0x4e69da]
frame #27: python() [0x4e69da]
frame #30: <unknown function> + 0x86242d (0x7f327aef642d in /localdisk/haozhe/pt-upstream/pytorch/torch/lib/libtorch_python.so)
frame #31: python() [0x4f80b3]
frame #33: python() [0x4e69da]
frame #36: python() [0x4e69da]
frame #40: python() [0x5c1157]
frame #41: python() [0x5bd170]
frame #42: python() [0x456423]
frame #46: <unknown function> + 0x29d90 (0x7f327cc6fd90 in /lib/x86_64-linux-gnu/libc.so.6)
frame #47: __libc_start_main + 0x80 (0x7f327cc6fe40 in /lib/x86_64-linux-gnu/libc.so.6)
frame #48: python() [0x58784e]
```
 - Second issue is we did not catch and rethrow exceptions in `omp parallel` region like [aten does](https://github.com/pytorch/pytorch/blob/2ad0e4197d1c6b2f0f732f2a6ec29de2af031c28/aten/src/ATen/ParallelOpenMP.h#L49-L51).
Without catching it, we will get
```
terminate called after throwing an instance of 'c10::Error'
terminate called after throwing an instance of 'c10::Error'
terminate called after throwing an instance of 'c10::Error'
terminate called recursively
[1]    2964411 IOT instruction (core dumped)
```
or just `terminate called recursively` with not mention to` c10::Error`. (since it is not catched)

**TestPlan:**
```
python test/inductor/test_cpu_repro.py -k test_catch_throw
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127868

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/common.py [hunk 1]
```diff
@@ -953,6 +953,47 @@ def ctx():
 
         return ctx()
 
+    def rethrow_exception(self):
+        @contextlib.contextmanager
+        def ctx():
+            self.writelines(
+                [
+                    "std::atomic_flag err_flag = ATOMIC_FLAG_INIT;",
+                    "std::exception_ptr eptr;",
```
Comment (jgong5) [inline torch/_inductor/codegen/common.py:962]: May I know why are we adding c++ specific code in the common buffer class?

File: torch/_inductor/codecache.py [hunk 2]
```diff
@@ -2483,6 +2484,7 @@ class CppPythonBindingsCodeCache(CppCodeCache):
                     [[unlikely]] throw std::runtime_error("requires %s args");
                 %s
             } catch(std::exception const& e) {
+                torch::translate_exception_to_python(std::current_exception());
```
Comment (jgong5) [inline torch/_inductor/codecache.py:2487]: how about ctypes binding?

**Review by jgong5 — CHANGES_REQUESTED** [review]
Tests failing. Can you add some brief description on how the problem is fixed?

File: torch/_inductor/codegen/common.py (hunk 1 above)
Comment (zhuhaozhe) [reply torch/_inductor/codegen/common.py]: Thanks for advice, I moved it to `cpp_utils.py`.

File: torch/_inductor/codegen/cpp.py [hunk 3]
```diff
@@ -3982,6 +3992,8 @@ def parallel(self, threads):
         if not self.in_parallel:
             self.num_threads = threads
             self.in_parallel = True
+            self.stack.enter_context(self.code.indent())
+            self.stack.enter_context(rethrow_exception(self.code))
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:3996]: Is it correct that we only rethrow exception when the loop is parallelized?

**Pull request closed without merging**